proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,280 @@
+blocks of non vote transactions. N number of builders can be scheudled
+concurrently to build blocks.
+
+While a leader is scheduled, they receive and encode votes as normal.
+Any well formed UserBlocks that were received from the previous or
+current UserBlockSlot are added to the leaders PoH ledger as
+UserBlockEntry.
+
+The N concurrent Builder create blocks out of user transactions.
+These are transmitted to the cluster via turbine.  The leader
+receives and decodes them and generates a UserBlockEntry, and adds
+it to PoH as soon as the leaders PoH has started the UserBlockSlot.
+
+Validators can vote on leader blocks by executing the votes, but
+before completing the execution of UserBlock transactions.
+
+### Fork Choice
+
+If a validator doesn't have the builder's UserBlock, the validator
+doesn't vote on the proposed block and tries to repair the UserBlock.
+That fork is ignored for fork choice.
+
+Otherwise the validator evaluates the forks as usual. 
+
+### UserBlock schedule
+
+There are 2 UserBlockSlots per slot. Builders produce 1 UserBlock
+per UserBlockSlot. So if the cluster is configured with 3 Builders,
+each leader slot would see a total of 6 UserBlocks per slot.  3 in
+the first half of the leaders slot, and 3 in the second half.
+
+The current leader schedule rotates each leader every 4 slots, but
+builders could be rotated at a different schedule then the leaders
+and rotation could be staggered.
+
+### UserBlock Compute Limits
+
+If the overall compute capacity for user transactions per leader
+block is 48m CU, and cluster is configured with 2 builders, then
+each UserBlock can use no more then 48m/4 or 12m CU.
+
+### UserBlock execution
+
+Execution of UserBlocks can happen asynchronously to voting.  When
+voting validators transmit their most recent BankHash, but it may
+be for an older parent slot then the slot they are voting on.
+
+Each UserBlock is assumed to have been created simultaneously during
+the UserBlockSlot that it was encoded by the leader.
+
+For each UserBlock in the PoH section spanning the UesrBlockSlot,
+the transactions are ordered by priority fee before execution.
+
+If two transactions from two different blocks have the same priority,
+they are ordered by which UserBlock appears first in the leaders
+PoH.
+
+If the leader's block is attached to the heaviest fork,
+the validator can start execution of the UserBlocks optimistically.
+Otherwise the validator should only execute UserBlocks on the
+heaviest fork.
+
+Duplicate transactions in the UserBlock are skipped over without
+any state changes. They must still be retained in the ledger because
+they have been hashed into the UserBlockEntry.
+
+### UserBlock format
+
+UserBlock follows the standard solana ledger format, except it
+MUST not contain any PoH ticks or votes.
+
+Invalid transactions are skipped, including duplicates, or those
+with invalid signatures, or completely malformed, and the rest of
+the block is processed.
```

Comment:
A bit confused on this part - why should we begin to allow a builder to include transactions without proper signatures?

Shouldn't the builder's be running sigverify and basic checks?

Reply:
Because if it is asynchronous, nodes may detect the invalid signature much later, well after they vote.  The only thing validators need to do to vote on forks is have the UserBlock data

Reply:
There is still an assumption that Builders validate transactions, making an effort to propose only UserBlocks with something that is valid, right?
As Builders are incentives for including transactions, they do not want to include garbage, and they could also be slashed for including transactions with invalid signatures or invalid in some other obvious way.
But, as pointed out, this should not stop the validator from including the UserBlock hash, even if it contains invalid transactions.